### PR TITLE
[MPS] Add `ctc_loss` backward pass

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #188187

Closes #160828

The backward pass is broken up into two kernel calls, one to calculate `log_beta` and one to perform logsumexp-reduction of `log_beta + log_alpha` and write the input grad.

Avoids some code duplication by refactoring the `log_alpha` code to be able to conditionally calculate `log_beta`.

The reduction step is parallelized by creating one thread per `T * N`, where `log_probs.shape = (T, N, C)`, `T` is the input length and `N` is the batch size. Each thread iterates over the target sequence, logsumexp-reducing the corresponding elements of `log_alpha` and `log_beta` into whichever element of the input gradient matches the target token. Then the thread iterates over its assigned row of the output grad and applies the formula $\nabla y = (y - \exp(R + L - \log y)) \nabla L$, where $\log y$ is `log_probs`, $R$ is the result of the reduction, $L$ is the loss calculated by the forward call.

[Notes on how ctc_loss works](https://github.com/kurtamohler/notes/blob/a56ce01e65ac591c6d851d2fa0d7438cbd523aa9/pytorch/ctc_loss.ipynb)

Co-authored with Claude

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/LossOps.metal [hunk 1]
```diff
@@ -103,38 +96,50 @@ kernel void ctc_loss(
...
     } else {
       target_token = params.BLANK;
       use_C = false;
     }
 
-    for (T_index t = 1; t < params.max_input_length; t++) {
+    for (T_index t_base = 1; t_base < params.max_input_length; t_base++) {
+      auto t = beta ? (params.max_input_length - (t_base + 1)) : t_base;
+      auto t_prev = beta ? (t + 1) : (t - 1);
       threadgroup_barrier(mem_flags::mem_device);
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/kernels/LossOps.metal:125]: We need to skip the boundary row in the beta sweeps or we will end up overwriting the initialization for it for padded inputs where `input_length < max_input_length` as it gets recomputed from the row above it that was set to -inf in the previous iteration.

```suggestion
      threadgroup_barrier(mem_flags::mem_device);
      if (beta && t == input_length - 1) continue;
```

A minimal test case that'll exercise the problematic path:

```
  def test_ctc_loss_backward_varlen_mps(self):
      # Backward must match CPU when input lengths are padded (shorter than T).
      # T=20, batch=2, 7 classes; the second sequence is only 11 steps long.
      T, N, C = 20, 2, 7
      log_probs = torch.randn(T, N, C).log_softmax(2)
      targets = torch.randint(1, C, (N, 4))
      input_lengths = [20, 11]   # 11 < T is what exposes the bug
      target_lengths = [4, 3]

      def input_grad(device):
          lp = log_probs.to(device).detach().requires_grad_()
          torch.nn.functional.ctc_loss(
              lp, targets.to(device), input_lengths, target_lengths
          ).backward()
          return lp.grad

      self.assertEqual(input_grad("cpu"), input_grad("mps").cpu())
```

**Review by jhavukainen — CHANGES_REQUESTED** [review]
There's a bug hiding in the beta sweep due to the boundary. Should be a small fix.

File: aten/src/ATen/native/mps/operations/LossOps.mm [hunk 2]
```diff
@@ -1297,7 +1299,18 @@ Tensor nll_loss2d_backward_mps(const Tensor& grad_output,
...
+    return "int32_t";
+  } else if constexpr (std::is_same_v<index_t, int64_t>) {
+    return "int64_t";
+  } else {
+    static_assert(false);
+  }
+}
+
+template <typename index_t, bool beta = false>
+static void ctc_loss_mps_kernel(std::optional<Tensor> loss,
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/LossOps.mm]: We can pass this by reference instead to align with the convention and make it a const while we are at it.
```suggestion
static void ctc_loss_mps_kernel(const std::optional<Tensor>& loss,
```

Comment (malfet) [reply aten/src/ATen/native/mps/operations/LossOps.mm]: I'm kind of curious if we can/should write an extra linter for file in say `native/mps` to do just that?

**Comment by malfet** [comment]
@claude review this and do you have a special skill about barriers?

File: aten/src/ATen/native/mps/kernels/LossOps.metal (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/LossOps.metal:125]: Thanks for catching that! I added the fix and figured I might as well add the test as well

**Comment by kurtamohler** [comment]
Thanks for the reviews, I've made the requested changes

**Review by jhavukainen — APPROVED** [review]

**Comment by jhavukainen** [comment]
Yeah we'll need one of the core maintainers / reviewers to approve these as well, my approvals are more of a "vote of confidence" than an actual approval I think. 😁 But regardless it looks good to me.

File: aten/src/ATen/native/mps/operations/LossOps.mm [hunk 3]
```diff
@@ -1297,7 +1299,18 @@ Tensor nll_loss2d_backward_mps(const Tensor& grad_output,
...
 template <typename index_t>
-static void ctc_loss_mps_kernel(Tensor& loss,
+std::string_view get_index_type_str() {
+  if constexpr (std::is_same_v<index_t, int32_t>) {
+    return "int32_t";
+  } else if constexpr (std::is_same_v<index_t, int64_t>) {
+    return "int64_t";
+  } else {
+    static_assert(false);
+  }
```
[the remark this answers is not in the document]
Comment (kurtamohler) [reply aten/src/ATen/native/mps/operations/LossOps.mm:1309]: Copilot is wrong about this. The static assert is dependent on `index_t` because of the `constexpr` conditions. Copilot claims that the assert will fail even for supported instantiations, but that isn't true, as the passing CI shows. I also tried adding a `get_index_type_str<float>();` call, just to double check that that raises an error at build time, and it does.

**Review by malfet — APPROVED** [review]
